File: projects/06/src/SymbolTable.java
import java.util.ArrayList;

public class SymbolTable {
    private final ArrayList<String> symbolList= new ArrayList<>();
    private final ArrayList<Integer> addressList = new ArrayList<>();

    public SymbolTable(){
        //initialize table here
        addEntry("R0", 0);
        addEntry("R1", 1);
        addEntry("R2", 2);
        addEntry("R3", 3);
        addEntry("R4", 4);
        addEntry("R5", 5);
        addEntry("R6", 6);
        addEntry("R7", 7);
        addEntry("R8", 8);
        addEntry("R9", 9);
        addEntry("R10", 10);
        addEntry("R11", 11);
        addEntry("R12", 12);
        addEntry("R13", 13);
        addEntry("R14", 14);
        addEntry("R15", 15);
        addEntry("SCREEN", 16384);
        addEntry("KBD", 24576);
        addEntry("SP", 0);
        addEntry("LCL", 1);
        addEntry("ARG", 2);
        addEntry("THIS", 3);
        addEntry("THAT", 4);
    }

    //Add a symbol and its given address
    public void addEntry(String symbol, int address){
        symbolList.add(symbol);
        addressList.add(address);
    }

    //checks for a given symbol in the list
    public boolean contains(String symbol){
        return symbolList.contains(symbol);
    }

    //Returns the address of a given symbol. Use only if symbolList contains symbol
    public int getAddress(String symbol){
        return addressList.get(symbolList.indexOf(symbol));
    }
}

File: projects/06/hack-assembler/src/symboltable.rs
use std::collections::HashMap;

// Defined as module constants so they can be shared.
// We store the static string references here.
const R_REGS: [&str; 16] = [
    "R0", "R1", "R2", "R3", "R4", "R5", "R6", "R7",
    "R8", "R9", "R10", "R11", "R12", "R13", "R14", "R15",
];

const PREDEFINED: [(&str, i32); 7] = [
    ("SCREEN", 16384),
    ("KBD", 24576),
    ("SP", 0),
    ("LCL", 1),
    ("ARG", 2),
    ("THIS", 3),
    ("THAT", 4),
];

pub struct SymbolTable {
    table: HashMap<String, i32>,
}

impl SymbolTable {
    pub fn new() -> Self {
        let mut table = HashMap::new();

        for (i, &name) in R_REGS.iter().enumerate() {
            table.insert(name.to_string(), i as i32);
        }

        for &(name, address) in &PREDEFINED {
            table.insert(name.to_string(), address);
        }

        SymbolTable { table }
    }

    pub fn add_entry(&mut self, symbol: &String, address: i32) {
        self.table.insert(symbol.to_string(), address);
    }

    pub fn contains(&self, symbol: &str) -> bool {
        let symbol = self.resolve_predefined(symbol).unwrap_or(symbol);
        self.table.contains_key(symbol)
    }

    pub fn get_address(&self, symbol: &str) -> i32 {
        // Returns the static ref if found, otherwise borrows the input 'symbol'
        let symbol = self.resolve_predefined(symbol).unwrap_or(symbol);
        *self.table.get(symbol).expect("Symbol not found in table")
    }

    // Checks incoming symbols so that we can me only predefined symbols case insensitive
    fn resolve_predefined(&self, symbol: &str) -> Option<&'static str> {
        for &(name, _) in &PREDEFINED {
            if symbol.eq_ignore_ascii_case(name) {
                return Some(name);
            }
        }

        if symbol.len() > 1 && symbol.len() <= 3 {
             let bytes = symbol.as_bytes();
             if bytes[0].to_ascii_uppercase() == b'R' {
                 // Try to parse the number part
                 if let Ok(idx) = symbol[1..].parse::<usize>() {
                     // Safe array access using the constant
                     if idx < R_REGS.len() {
                         return Some(R_REGS[idx]);
                     }
                 }
             }
        }

        None
    }
}
